### PR TITLE
DPL-1024-2 PBMC donor pooling - ancestor plates

### PR DESCRIPTION
**Pull request by yoldas**
Closes #

#### Changes proposed in this pull request

Updates for using multiple ancestor plates in driver file download

Summary:

The ancestor_purpose setting of an export configuration will make an additional controller instance variable`@ancestor_plate_list` populated with all ancestor plates of that purpose and available in the view template.

Background:

The existing export controller allows making an ancestor plate available to the export view template by specifying the ancestor plate purpose in the export configuration. This is useful in several cases, for example, access to parent plate barcode and relating source wells specified in transfer templates in driver file exports. However, only one ancestor plate is made available to the view, the first, which assumes a single parent for the current plate being viewed and download requested.

In cases where a plate has (or may have) multiple ancestors of a specific purpose, such as pooling from multiple source plates, as in scRNA donor pooling, we need access to more than one parent in the view to interpret the transfer request and generate a column with parent barcode. This PR enables that.

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: app/controllers/exports_controller.rb [hunk 1]
```diff
@@ -43,16 +44,28 @@ def locate_labware
...
 
+  # Returns an array of all ancestor plates of @plate that match a specific
+  # purpose (defined in the export configuration). If no such plates are found,
+  # an empty array is returned. The result array is made available to the view
+  # that generates the export as @ancestor_plate_list by the show method.
+  #
+  # @return [Array, Sequencescape::Api::V2::Plate] An array of Plate records if
+  #   any are found, otherwise an empty array.
+  def locate_ancestor_plate_list
+    return nil if export.ancestor_purpose.blank?
```
Comment (andrewsparkes) [inline app/controllers/exports_controller.rb:61]: should that return empty array instead of nil to be consistent?

File: app/controllers/exports_controller.rb [hunk 2]
```diff
@@ -43,16 +44,28 @@ def locate_labware
...
   end
 
+  # Returns an array of all ancestor plates of @plate that match a specific
+  # purpose (defined in the export configuration). If no such plates are found,
+  # an empty array is returned. The result array is made available to the view
+  # that generates the export as @ancestor_plate_list by the show method.
+  #
+  # @return [Array, Sequencescape::Api::V2::Plate] An array of Plate records if
+  #   any are found, otherwise an empty array.
+  def locate_ancestor_plate_list
```
Comment (andrewsparkes) [inline app/controllers/exports_controller.rb:60]: Please add an rspec test for this method.
I think it's useful and will be used in other pipelines.

**Review by andrewsparkes — CHANGES_REQUESTED** [review]
Looks good, just add a test and see comment about nil / empty array.

**Comment by KatyTaylor** [comment]
Could you add a bit more to the PR description, explaining what the problem was that you're fixing? Thanks

File: app/controllers/exports_controller.rb [hunk 3]
```diff
@@ -43,16 +44,28 @@ def locate_labware
...
+    return nil if export.ancestor_purpose.blank?
+
+    # Collect plate ids from Asset resource results to fetch Plates later.
```
Comment (KatyTaylor) [inline app/controllers/exports_controller.rb:63]: Can we avoid using the word Asset? 
1. Isn't @plate actually coming from a V2 Plate API query in `locate_labware`?
2. Asset is a historical term from before it was split into Labware and Receptacle, and I'd like to get rid of all references to the word 'Asset' from SS & Limber to avoid confusion.

File: app/controllers/exports_controller.rb (hunk 1 above)
Comment (yoldas) [reply app/controllers/exports_controller.rb]: Thanks for spotting that. Yes, it should return an empty array as described in the method documentation.

File: app/controllers/exports_controller.rb (hunk 2 above)
Comment (yoldas) [reply app/controllers/exports_controller.rb:60]: I have added test for the behaviour of the controller when dealing with multiple ancestor plates using this method and configuration.  It verifies that @ancestor_plate_list is correctly assigned and the controller makes it available to the view template and the response body matches the expected output, in two contexts when it is configured and not configured.

**Comment by yoldas** [comment]
> Could you add a bit more to the PR description, explaining what the problem was that you're fixing? Thanks

I have updated the PR description to explain the problem being fixed.

File: app/controllers/exports_controller.rb (hunk 3 above)
Comment (yoldas) [reply app/controllers/exports_controller.rb]: I have updated the method documentation to make it clear.

**Review by KatyTaylor — APPROVED** [review]
Thanks for the more detailed description
